bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/CreateNewLogTest.java

```diff
@@ -142,6 +151,237 @@ public void testCreateNewLogWithNoWritableLedgerDirs() throws Exception {
         assertTrue("Wrong log id", entryLogManager.getCurrentLogId() > 1);
     }
 
+    /*
+     * entryLogPerLedger is enabled and various scenarios of entrylogcreation are tested
+     */
+    @Test
+    public void testEntryLogPerLedgerCreationWithPreAllocation() throws Exception {
+        /*
+         * I wish I could shorten this testcase or split it into multiple testcases,
+         * but I want to cover a scenario and it requires multiple operations in
+         * sequence and validations along the way. Please bear with the length of this
+         * testcase, I added as many comments as I can to simplify it.
+         */
+
+        ServerConfiguration conf = TestBKConfiguration.newServerConfiguration();
+
+        // Creating a new configuration with a number of ledger directories.
+        conf.setLedgerDirNames(ledgerDirs);
+        conf.setIsForceGCAllowWhenNoSpace(true);
+        // preAllocation is Enabled
+        conf.setEntryLogFilePreAllocationEnabled(true);
+        conf.setEntryLogPerLedgerEnabled(true);
+        LedgerDirsManager ledgerDirsManager = new LedgerDirsManager(conf, conf.getLedgerDirs(),
+                new DiskChecker(conf.getDiskUsageThreshold(), conf.getDiskUsageWarnThreshold()));
+        EntryLogger entryLogger = new EntryLogger(conf, ledgerDirsManager);
+        EntryLoggerAllocator entryLoggerAllocator = entryLogger.entryLoggerAllocator;
+        EntryLogManagerForEntryLogPerLedger entryLogManager = (EntryLogManagerForEntryLogPerLedger) entryLogger
+                .getEntryLogManager();
+
+        /*
+         * no entrylog will be created during initialization
+         */
+        int expectedPreAllocatedLogID = -1;
+        Assert.assertEquals("PreallocatedlogId after initialization of Entrylogger",
+                expectedPreAllocatedLogID, entryLoggerAllocator.getPreallocatedLogId());
+
+        int numOfLedgers = 6;
+
+        for (long i = 0; i < numOfLedgers; i++) {
+            /* since we are starting creation of new ledgers, entrylogid will be ledgerid */
+            entryLogManager.createNewLog(i);
+        }
+
+        Thread.sleep(100);
```

Comment:
why do you need sleep here?
if createNewLog() is async please consider something more predictable/less error prone. i.e. return future you can wait on.

Reply:
guava to the rescue! removed these sleeps, for entryloggerallocator's allocatorexecutor using newDirectExecutorService()